### ISSUE TITLE
[RFC FS-1001 Discussion] String Interpolation

### ISSUE THREAD

**Issue by v2m** [fsharp/fslang-design#6]
This issue is used to track discussions of [F# RFC FS-1001 - "String interpolation"](https://github.com/fsharp/FSharpLangDesign/blob/master/RFCs/FS-1001-StringInterpolation.md)

**Comment by forki** [comment]
Unfortunately the corresponding pull request on codeplex was closed because of release pressure. Can we somehow revive this? String interpolation is a feature that a lot of users would like to see.

**Comment by forki** [comment]
Ok I rebased it on master: https://github.com/Microsoft/visualfsharp/pull/921

**Comment by enricosada** [comment]
@dsyme can you or someone with permission fix broken link in pr description? now is https://github.com/fsharp/FSharpLangDesign/blob/master/FSharp-4.1-Proposals/StringInterpolation.md

**Comment by forki** [comment]
done

**Comment by enricosada** [comment]
a question: are we supporting localization correctly?

[msdn about c# string interpolation](https://msdn.microsoft.com/en-us/library/dn961160.aspx)

It's not only about a better string format, need to support i18n too. C# does that with [FormattableString](https://msdn.microsoft.com/en-us/library/system.formattablestring.aspx)

```
var product = "Raspberry Pi";
var price = 25.0;
Console.WriteLine(LocalizationHelper.Format(
    $"The price of {product} is {price}.";
));
```

> However, the string interpolation will be compiled to a FormattableString with “The price of {0} is {1:C}” as the format that will get passed to GetLocalized. In other words, the people doing the localization will still have to localize with numeric indices instead of the more readable named tokens

**Comment by enricosada** [comment]
@forki not your pr, this pr description from @v2m

**Comment by jarlestabell** [comment]
(I posted this originally in the wrong thread (the implementation thread).)

I assume most people here know the Scala solution to this, but just in case I want to point that one out, as I think it is quite elegant.
I assume it is drastically much heavier to implement though and it may overlap a bit with type provider territory.

The key point is that it has a concise syntax and is extensible.
So the same concept can be used both for string interpolation without formatting:

    val name = "James"
    println(s"Hello, $name")  // Hello, James

with formatting:

    val height = 1.9d
    val name = "James"
    println(f"$name%s is $height%2.2f meters tall")  // James is 1.90 meters tall

The above examples are taken from:
http://docs.scala-lang.org/overviews/core/string-interpolation.html


And plain SQL queries: (http://slick.typesafe.com/doc/3.1.1/sql.html#string-interpolation)

    def insert(c: Coffee): DBIO[Int] =
     sqlu"insert into coffees values (${c.name}, ${c.supID}, ${c.price}, ${c.sales}, ${c.total})"

and even things like ReactJS-like JSX/TSX xml-based syntax:

    xml”””
	      <body>
    		<a href = “some link”> ${linktext} </a>
	    </body>
	    ”””

(example from http://docs.scala-lang.org/sips/completed/string-interpolation.html)

**Comment by enricosada** [comment]
scala syntax **and extensibility** it's really interesting, `StringContext` it's like [System.FormattableString](https://msdn.microsoft.com/en-us/library/system.formattablestring(v=vs.110).aspx) and the rest it's like `sprintf`.
The extensibility is really nice to have to make a `FormattableString` version (useful for localization)

**Comment by jarlestabell** [comment]
Note that ES6 seems to have the same syntax as Scala, even down to the extensibility. ("tagged template literals")
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Template_literals
When working in F#, Scala's string interpolation is one of the few things I miss from Scala.
(Although I have done type mistakes with doing string interpolation in Scala)

**Comment by cloudRoutine** [comment]
I'm with @isaacabraham on using the `"%{expr}"` syntax over parentheses. Especially if they allow arbitrary expressions, then it is syntactically coherent with the F# conventions of using `%` for substitution and `{ }` to delineate a block of scope.

Since we'll know the type the expression resolves to we can determine which formatting specifications to enable. For most types -
* `%10{**expr**}` - sets the width of the embedded expression
* `%*{**expr**}` - sets the width dynamically  
* `%-{**expr**}` - left justifies the embedded expression

If it resolves to a numeric type
* `%0{**numeric-expr**}` - pads with zeros
* `%+{**numeric-expr**}` - shows a plus sign
* `% {**numeric-expr**}` - shows a blank in place of a plus sign
* `%.5{**numeric-expr**}` - control precision when expr is a float 

The primitive types can follow the same style of formatting that `printf` uses, defaulting to

`%b`, `%c`, `%i`, `%u`, `%x`, `%o`, `%f`, `%M`,

Additional Questions -
- Is interpolation allowed in every string, only `printf` strings or will it need a prefix operator?  
- Should interpolation be allowed in bare strings if only interpolations substitution is used? 
   e.g. `"hello %{name}, I have your %10{item}"`

- Would this be valid `"%{root}"B` ?
- Should there be an additional specifier to indicate `%O` format instead of `%A` format, or vice versa depending on whichever becomes the default.

**Comment by smoothdeveloper** [comment]
@cloudRoutine sorry for being late in the party, but you say

> and `{ }` to delineate a block of scope.

Would it be conflicting to use `( )` instead? it seems what we want to interpolate with is F# expressions, and parens are used to delineate those more often than curly braces which are only used for records, I think this would be more consistent and would look less strange if one decides to interpolate with an inline record expression or object expression (having `%{{field1=1}}` would look kind of strange).

**Comment by cloudRoutine** [comment]
@smoothdeveloper `{  }` are used for record expressions, in constructor overloads, object expressions, seq expressions, query expressions, and computation expressions. In each of these cases the curly braces demarcate a context with slightly different semantics than the scope that encompasses it - through context specific keywords, field assignments, member and interface definitions, or custom operations.

`( )` [are used to indicate the order in which expressions are evaluated, to delimit a tuple](https://docs.microsoft.com/en-us/dotnet/articles/fsharp/language-reference/symbol-and-operator-reference/#table-of-symbols-and-operators), to use an active pattern or operator like a function, or to represent unit.  Whereas `{  }` are always used to define a distinct scope.

`"  %{{field1=1}}"` - declaring an anonymous record expression in the middle of a string interpolation seems like a very unlikely use case. 

`"  %((20.M, 10.0, 5))"` - a tuple substitution seems just as if not more probable than your "strange" case if parens were used as the delimiters.

And if for some bizarre reason you feel the need to define a bunch of expressions within the scope of the substitution block, with several lambdas and function calls, there'll be parens all over the place. But it'll be nice and easy to see the bounds around it, thanks to those `{  }` 😜

**Comment by smoothdeveloper** [comment]
@cloudRoutine thanks, it makes more sense with the expanded description, I took scope as something akin to how it is used in C# but not as being different context as you explain.

As for what goes in the interpolated strings, you'd be surprised of what can be found in code in the wild even though that feature was only recently added to C# 6, I won't be surprised to find tuples or records, and why not object expression since it can be done :)

**Comment by cloudRoutine** [comment]
By adding another `printf` format specifier to use with interpolateds string we could add an even stronger degree of type safety.

instead of needing to write helper functions like
```
 let makeString (arg:MyDu) =
    sprintf "adjsk -- %O" arg
```
the `%T` specifier could be used with an interpolated insertion that only accepts types

` sprintf "adjsk -- %T{MyDu}" MyDu.Case1`

which may not immediately seem useful, but for partial application it could provide a lot of value.

perhaps `%t` could use ToString like `%O` and `%T` StructuredFormatDisplay like `%A`?

**Comment by vasily-kirichenko** [comment]
I think we are in a classical "analysis paralysis" :) What about implementing a minimal functionality first, keeping the door open for further improvements? 

* choose the C# syntax as is: `$"{ }`
* no Scala-like extensions
* no specifiers at all
* F# types are formatted with `%+A`, everything else - with `%O`

So we would ended up with this:

```fsharp
let s = $"This feature is {x}!" "awesome"
```

Great things about it:

* no need to switch your mind between two syntaxes when working on C# - F# solutions
* no need for `sprintf`, which looks old fashioned compared to C# and other langs implementations
* specifiers and extensions can be added later, no breaking changes
* quite easy to implement? 

In short, I believe we should implement _anything_ to move forward with this, then, according to feedback, make it great.

**Comment by jhurdlow** [comment]
I'm with vasily here... Go with the existing C# syntax. It's clean and it works. Most importantly it will be a natural transition for those of us who have to bounce back and forth between C# and F# all day. I find myself trying to do this all the time in F#, then remembering it's (sadly) not there (yet). If you need custom formatting, just do it inside the {}, after all it's just an expression. This is the feature I miss most from C# when working in F#, so it would be nice to see this feature sooner rather than later.

**Comment by realvictorprm** [comment]
I really would like to have this feature soon in F# with the same syntax it's implemented in C# because I also think it works very well!

Is there anything we can do to continue the discussion and or to receive an approval?

**Comment by dsyme** [comment]
@realvictorprm Yes, we need to find a way to progress this.  We need to go back to some basics - e.g. I'm now inclined to agree we should go with the C# syntax for this

**Comment by realvictorprm** [comment]
Hm as soon as it's approved I would love to try implementing it to get a grasp of how adding features to the compiler works.

**Comment by dsyme** [comment]
@realvictorprm Adding string interpolation is "approved in principle" (i.e. "planned" in https://fslang.uservoice.com/forums/245727-f-language/suggestions/6002107-steal-nice-println-syntax-from-swift)

Next step is either to implement the C# syntax, or edit the RFC to capture that as a proposed design

**Comment by realvictorprm** [comment]
So I could just start looking how to start implementing it, right?

I've a pretty big interest in this because generator stuff etc. involves big string interpolation. On the C# side the code is very readable thanks to the interpolation feature but with F# readability is lacking here really much.

Is there any pointer you can give me where to code should start? E.g the place where strings are being checked?

**Comment by vasily-kirichenko** [comment]
@realvictorprm see initial attempt https://github.com/Microsoft/visualfsharp/pull/921

**Comment by realvictorprm** [comment]
Super awesome, many thanks!

**Comment by realvictorprm** [comment]
@dsyme I require a bit assistance because I already spend about 5 hours on this without much success. Before I'll Try to explain my plan /specification.

The implementation should ensure that any string which starts with a dollar symbol can be interpolated with the following syntax: `$"{ExpressionOrConstant}".
The given expression or constant is formatted like it would have been formated with the `%A` specifier. 

To sum this up, it should work without printf or sprintf which requires a new treatment in the lexer and parser. Most of the time I spend with understanding what's happening and waiting till the compiler is compiled(I should open an issue for that...). My preferred approach would be (partly what I already tried to do)
1. Adding a new token which is being created if a dollar symbol occurs (called STRING_INTERPOLATED)
2. This indicates that the content of the string must be validated in the lexer or parser
3. The parser maps this to a special AST type called again string interpolated or something else.
4. The type check is done afterwards in the sections where formatting strings are type checked just like in @forki''s PR (grabbing the local environment and ensuring that all specified values exist).
5. Then an operation is added which formats the string at runtime (should be similar to the printf mechanics I guess)

If my thoughts are wrong or if I miss something please tell me as soon as possible.

**Comment by realvictorprm** [comment]
PS: Please bear with me. It's the first time I touch a compiler in its core and I'm still learning how all mechanisms there work.

**Comment by mrange** [comment]
If you plan to have STRING_INTERPOLATED as a specially tagged string that will later be processed you should be able to reuse most of the actual string parser code.

**Comment by mrange** [comment]
`let startString` on line 122 IIRC.

**Comment by realvictorprm** [comment]
@mrange The milestones would be for me:
1. Value interpolation support:
1.1 Get interpolated string identification to the type checkers.
1.2 Apply simple validation of the interpolated string in the lexer / parser. 
1.3 Apply typechecking.
2. Discuss whether expression support is preferred too.

**Comment by realvictorprm** [comment]
Good to know that I'm working on the right locations, very good to know @mrange.

**Comment by mrange** [comment]
I am not sure if this works but I was thinking perhaps like this:

l122: `let startString stringToken args (lexbuf: UnicodeLexing.Lexbuf) =`
l137: `stringToken (Lexhelp.stringBufferAsString s))`

All places `startString` is used today add `STRING` like so:
`startString STRING args lexbuf`

For interpolated strings perhaps this:
```
 | '$' '"' 
     { let buf,fin,m = startString STRING_INTERPOLATED args lexbuf 
       if not skip then (STRING_TEXT (LexCont.String(!args.ifdefStack,m))) else string (buf,fin,m,args) skip lexbuf }
```

**Comment by mrange** [comment]
Hopefully I am not leading you in the wrong direction here so look at my suggestions with a critical eye.

**Comment by realvictorprm** [comment]
So far I see you're mostly telling me for what I needed 6 hours to figure out.

**Comment by realvictorprm** [comment]
@mrange so far I see I must map the interpolated string to a specific **new AST type**, right?

**Comment by mrange** [comment]
Not a new AST type I guess but a new union case in the existing AST type.

**Comment by realvictorprm** [comment]
Ups yeah that's what I meant for sure ;)

**Comment by mrange** [comment]
And a new case:
`pars.fsy, L2657`: `  | STRING_INTERPOLATED { SynConst.StringInterpolated $1 }`

`StringInterpolated` is your new AST case.

**Comment by realvictorprm** [comment]
yeah I'm a bit before this point, I just added all necessary stuff in the parser and lexer. Now after I understand how it works it's a lot easier 😅

**Comment by realvictorprm** [comment]
Thank you very much for your advice @mrange. I managed to get the information into the type checker.

**The new PR for C# like String-Interpolation is located [here](https://github.com/Microsoft/visualfsharp/pull/3593)!**

**Comment by dsyme** [comment]
@realvictorprm Great work!

**Comment by dmitry-a-morozov** [comment]
I cannot find here what's the final decision on ```FormattableString``` implicit conversion but I would like to point out to this api
https://docs.microsoft.com/en-us/dotnet/api/microsoft.entityframeworkcore.relationalqueryableextensions.fromsql?view=efcore-2.0#Microsoft_EntityFrameworkCore_RelationalQueryableExtensions_FromSql__1_System_Linq_IQueryable___0__System_FormattableString_
And related discussion:
https://github.com/aspnet/EntityFrameworkCore/issues/9734#issuecomment-328243741
https://github.com/aspnet/EntityFrameworkCore/issues/10080.

Not defending EF Core team decision but reality is more APIs like that will show up and it would be nice to have smooth inter-op in F#.

**Comment by realvictorprm** [comment]
I didn't saw that one @dmitry-a-morozov.

I just went through it a small bit and it sounds like a formatting function or whatever. Just something which wraps the operation which would be done instantly. However as always I highly discourage any and I really mean _any_ implicit conversion in F#. Therefore if we want interop with other stuff out there we must acquire it with either 1. a different syntax or 2. we change our mind about interpolated strings in such a way that they are functions which can be executed (which is very unlikely).

I'm pretty open for using another prefix for strings to signal that the formatting shouldn't be executed instantly rather instead a function or formatting object should be returned (maybe it's a function string then 😛 ?). This however would be the same as shortcutting the sprintf function @dmitry-a-morozov I think. Maybe something like this (prefix stands for function)?
```fsharp
let formatDefinition = f"let {name} = {definition}"
```

**Comment by iddan** [comment]
What is the status of this?

**Comment by cartermp** [comment]
@iddan There have been a few PRs from the community, but none have been merged.

**Comment by iddan** [comment]
What blocks them from being merged?

**Comment by cartermp** [comment]
We hold a high bar for features getting merged into the compiler, and that means a significant time investment on behalf of us and the person implementing it. We haven't prioritized the feature above other things to do ourselves, so that time investment lies on a community member if they have a strong desire to see the feature in an upcoming version of the language.

**Comment by iddan** [comment]
Is there a transparent lifecycle of a proposal?
For example: [ECMAScript Proposals](https://github.com/tc39/proposals), [Rust rfcbot](https://rfcbot.rs/)

**Comment by enricosada** [comment]
There are info in the [README](https://github.com/fsharp/fslang-design/blob/master/README.md)

More info in the RFC doc https://github.com/fsharp/fslang-design/blob/master/RFCs/FS-1001-StringInterpolation.md

**Comment by yatli** [comment]
according to https://github.com/Microsoft/visualfsharp/pull/3593, there are currently two work items:

1. test cases
2. recursive parsing on nested interpolation

2) shall be also covered in 1). So effectively there's one single thing to do -- some TDD challenge/response iterations :)

I suggest, if we care enough about this feature, we can at least "crowdsource" some test cases.

**Comment by yatli** [comment]
I'm trying to bring it back upon current `master`.

**Comment by yatli** [comment]
creating a temp PR here: https://github.com/yatli/visualfsharp/pull/1

**Comment by yatli** [comment]
I'd suggest decompose FS-1001 into two components, and make them more native to what we already have by using simple transformations of the syntax tree rather than introducing new mechanisms:

Part 1: transform `$"some string"` to `(sprintf "some string")`
Part 2: transform `"some %{interpolated} string"` to ` "some %A string" interpolated` -- the formatting options can also be given inline, for example, transform `"some %{s:interpolated} string"` should be transformed to: `"some %s string" interpolated`.

In this way string interpolation will work well with the `printf` family.

I'll have to think about mixing inline arguments with plain formatting placeholders -- the order is the key, for example: ` $"this is the %d-th %{s:interpolated} string" nr_str` -- when transforming the AST we have to take the poisition of `nr_str` into consideration, or make it two-step formatting (which brings some overhead though)

**Comment by yatli** [comment]
Actually part 1 can be done in the library not the compiler, by making $ an operator that works just like sprintf.

**Comment by yatli** [comment]
After some thoughts, i think it’s impossible to have one-level formatting if inline arg and placeholders are mixed. The correct way of handling this is to create a closure that captures the inline argument, and output a new formatter.

**Comment by yatli** [comment]
I'm opening a draft PR, and I will begin documenting my design.

**Comment by Alex2357** [comment]
Completely agree with @vasily-kirichenko 

> I think we are in a classical "analysis paralysis" :) What about implementing **a minimal functionality first**, keeping the door open for further improvements?


If this will work
```csharp
var s = $"hello, {name}"
```
 it would cover 85-95% of my problems. For the remaining ones I can write sprintf. I believe I'm not alone and most of the people would be quite happy just having this.

**Comment by realvictorprm** [comment]
It's not like I've done that in the past lol. Even if I started it only :D

**Comment by dsyme** [comment]
After far too long I've added a comment about the design for this RFC here: https://github.com/dotnet/fsharp/pull/6770#issuecomment-507268783

**Comment by realvictorprm** [comment]
Thanks :)!

Don Syme <notifications@github.com> schrieb am Mo., 1. Juli 2019, 15:40:

**Comment by dsyme** [comment]
I've rewritten the RFC. Closing in favour of new discussion thread https://github.com/fsharp/fslang-design/issues/368
